adding option to return kinematic equations implicitly in KanesMethod #22818

zouhairm opened the pull request:
Fixes #22626 

This adds an option to KanesMethod to expose kinematic matrices in implicit form.

Default behavior should remain unchanged, so existing tests should be fine. Added one test to exercise the new code.

Closing PR https://github.com/sympy/sympy/pull/22818 to take a different approach as proposed by @moorepants in https://github.com/sympy/sympy/pull/22818#issuecomment-1086811861 . Doing this in a different PR in case we decide to pursue the other method at some point (and because I have some 3rd party code relying on that branch right now which I'll migrate to this PR once it's approved)

<!-- BEGIN RELEASE NOTES -->
* physics.mechanics
  * Added option to return kinematics and full mass matrix in implicit form in KanesMethod

<!-- END RELEASE NOTES -->

zouhairm committed:
redo of #22626 based on feedback

zouhairm commented:
@moorepants - plz let me know if this along the lines of what you're looking for.

zouhairm committed:
skipping test error on symengine

moorepants commented:
@TJStienstra would you like to review this? It would be a good practice for your GSoC.

TJStienstra commented:
Yeah fine, thanks for showing. Will have a lot more time this week, so will look into it somewhere in the coming days (probably tomorrow).

zouhairm commented:
thanks for the comments @TJStienstra - will try to address them this week.

zouhairm committed:
addressing reviewers comments

zouhairm commented on sympy/physics/mechanics/tests/test_kane.py:
```diff
@@ -348,3 +353,113 @@ def test_input_format():
+        [x.count_ops() for x in KM.forcing_full_implicit] +
+        [x.count_ops() for x in KM.mass_matrix_full_implicit]
+    )
+    assert n_ops_implicit / n_ops_explicit < .05
```
the other test is actually in `test_two_dof`

I added another test, although turned out this is not so easy as I couldn't use the whole explicit form as a way to check things against since the explicit equations are too slow to handle (hence the whole point of this PR :p)

I've  basically used the explicit form for the linear velocities which are 'trivial', and a 'textbook' solution for the quaternion rates which are more involved and thus their explicit form is too complex.

zouhairm commented:
@TJStienstra - sorry for the much longer than expected delay as some higher priorities ended up taking over.

I believe I've addressed all of your comments. I'm hoping we can get this merged as it'd be great to get our internal codebase updated to use a 'built wheel' from sympy as opposed to having to build it from scratch from my branch.

Thanks!

TJStienstra commented:
Nice work, like the tests and I agree it would be nice to have this in the code base as soon as possible. Overall there are some small typos, but will be addressing a more general refactoring of `KanesMethod` in #24093.

However just discussed the PR with @moorepants and @brocksam and came with the following proposal to reduce the current polutting (~100 lines of code for these properties and 3 extra external methods and 3 protected methods). By implementing an attribute `explicit`, which is defaultly set to `True` we can drastically reduce it. Here is some example code:

```py
kanes = KanesMethod(...)
kanes.kanes_equations()
kanes.forcing_kin  # Explicit kinematic forcing vector
kanes.mass_matrix_full # Explicit full mass matrix
kanes.explicit = False
kanes.forcing_kin  # Implicit kinematic forcing vector
```

@zouhairm What do you think of this proposal (it is in the end a bit of what you've started with)?

Will be giving the entire suggestion, that you can check and commit, in a few minutes.

TJStienstra commented on sympy/physics/mechanics/tests/test_kane.py:
```diff
@@ -77,6 +79,10 @@ def test_two_dof():
     assert simplify(KM.rhs() -
                     KM.mass_matrix_full.LUsolve(KM.forcing_full)) == zeros(4, 1)
 
+    # Check that the implicit form also gives the same result
+    assert simplify(KM.rhs() -
+                    KM.mass_matrix_full_implicit.LUsolve(KM.forcing_full_implicit)) == zeros(4, 1)
```
```sugggestion
    rhs = KM.rhs()
    assert simplify(rhs -
                    KM.mass_matrix_full.LUsolve(KM.forcing_full)) == zeros(4, 1)

    # Check that the implicit form also gives the same result
    KM.explicit = False
    assert simplify(rhs -
                    KM.mass_matrix_full.LUsolve(KM.forcing_full)) == zeros(4, 1)
```

TJStienstra commented on sympy/physics/mechanics/tests/test_kane.py:
```diff
@@ -348,3 +354,147 @@ def test_input_format():
+        [x.count_ops() for x in KM.forcing_full] +
+        [x.count_ops() for x in KM.mass_matrix_full]
+    )
+    n_ops_implicit = sum(
+        [x.count_ops() for x in KM.forcing_full_implicit] +
+        [x.count_ops() for x in KM.mass_matrix_full_implicit]
+    )
```
```suggestion
    KM.explicit = False
    n_ops_implicit = sum(
        [x.count_ops() for x in KM.forcing_full] +
        [x.count_ops() for x in KM.mass_matrix_full]
    )
```

TJStienstra commented on sympy/physics/mechanics/tests/test_kane.py:
```diff
@@ -348,3 +354,147 @@ def test_input_format():
+    # Instead, we check that the kinematic equations are correct using more fundamental tests:
+    #
+    # (1) that we recover the kinematic equations we have provided
+    assert (KM.mass_matrix_kin_implicit * KM.q.diff() - KM.forcing_kin_implicit) == Matrix(kinematic_eqs)
```
```suggestion
assert (KM.mass_matrix_kin * KM.q.diff() - KM.forcing_kin) == Matrix(kinematic_eqs)
```

TJStienstra commented on sympy/physics/mechanics/tests/test_kane.py:
```diff
@@ -348,3 +354,147 @@ def test_input_format():
+    # (2) that rate of quaternions matches what 'textbook' solutions give
+    # Note that we just use the explicit kinematics for the linear velocities
+    # as they are not as complicated as the angular ones
+    qdot_candidate = KM.forcing_kin
```
```suggestion
    KM.explicit = True
    qdot_candidate = KM.forcing_kin
```

TJStienstra commented on sympy/physics/mechanics/tests/test_kane.py:
```diff
@@ -348,3 +354,147 @@ def test_input_format():
+
+    # sub the config constraint in the candidate solution and compare to the implicit rhs
+    lambda_0_sol = solve(config_cons[0], q_att_vec[0])[1]
+    lhs_candidate = simplify(KM.mass_matrix_kin_implicit * qdot_candidate).subs({q_att_vec[0]: lambda_0_sol})
+    assert lhs_candidate == KM.forcing_kin_implicit
```
```suggestion
    KM.explicit = False
    lhs_candidate = simplify(KM.mass_matrix_kin * qdot_candidate).subs({q_att_vec[0]: lambda_0_sol})
    assert lhs_candidate == KM.forcing_kin
```

tjstienstra reviewed (commented):
I think I've added all change suggestion. Only in the `__init__` you still have to add `self.explicit = True` as last line.

zouhairm commented:
The suggestion sounds reasonable to me (In fact that was the initial [implementation](https://github.com/sympy/sympy/pull/22818#discussion_r836788688) but we moved away from it)

--
Thanks for taking the time to type up the suggestions, unfortunately, github is not allowing me to commit them (due to this issue https://github.com/isaacs/github/issues/1755)

I can manually type them up, but if you happen to have this in branch already, can you either submit a PR to my fork, or perhaps post a patch file?

TJStienstra commented:
Almost have a patch file if you'd like

TJStienstra commented:
[explicit_attribute.zip](https://github.com/sympy/sympy/files/9677135/explicit_attribute.zip)

TJStienstra commented:
Using `explicit_kinematics` as attribute is indeed more clear.

zouhairm committed:
implementing reviewers suggestion of an explicit flag

zouhairm committed:
fixing code quality due to \m math in comment block

zouhairm committed:
removing copy to support SYM_ENGINE

TJStienstra commented on sympy/physics/mechanics/tests/test_kane.py:
```diff
@@ -74,8 +78,31 @@ def test_two_dof():
 
-    assert simplify(KM.rhs() -
-                    KM.mass_matrix_full.LUsolve(KM.forcing_full)) == zeros(4, 1)
+    # Check that the explicit form is the default and kinematic mass matrix is identity
+    assert KM.explicit_kinematics
+    assert KM.mass_matrix_kin == eye(2)
+
+    # Check that the implicit and explicit matrices are different when involving kinematics
+    mf_keys = ['mass_matrix', 'mass_matrix_kin', 'mass_matrix_full',
+               'forcing', 'forcing_kin', 'forcing_full']
+    KM_matrices = {}
+    for explicit_kinematics in [False, True]:
+        KM.explicit_kinematics = explicit_kinematics
+        KM_matrices['explicit' if explicit_kinematics else 'implicit'] = {k:KM.__getattribute__(k) for k in mf_keys}
+
+    for k in mf_keys:
+        if k in ['mass_matrix', 'forcing']:
+            # dynamic matrix/vector are unchanged whether we use explicit or implicit kinematics
+            assert KM_matrices['explicit'][k] == KM_matrices['implicit'][k]
+        else:
+            # but matrix/vector involving kinematics are expected to be different
+            assert KM_matrices['explicit'][k] != KM_matrices['implicit'][k]
+
+    # Check that whether using implicit or explicit kinematics the RHS equations are consisten with the matrix form
+    for explicit_kinematics in [False, True]:
+        KM.explicit_kinematics = explicit_kinematics
+        assert simplify(KM.rhs() -
+                        KM.mass_matrix_full.LUsolve(KM.forcing_full)) == zeros(4, 1)
```
I would personally not usethese kind of test loops, since they are not nice to debug. Rather write them out, since you don't even use less lines in the end. Besides that I would also suggest to directly check the implicit kinematic mass matrix and forcing vector (they are really simple), i.e. `KM.mass_matrix_kin == Matrix([[1/2, 0], [0, 2]])` (is also bit more useful and accurate)

P.S. Try to keep the lines within 80 characters

moorepants replied:
Let's just try to get this in. We've nitpicked @zouhairm enough :)

zouhairm replied:
- the for loop was on purpose, not to save lines of code, but to make it clear that the only thing changing between the two tests is the explicit flag. pytest should print out the variables when something fails, so hopefully shouldn't be too difficult to debug.

-  I've removed one of them anyway by implementing a more, excuse the pun, explicit matrix check as you suggested as opposed to just checking that they are different

- sorry about the 80 chars: I know it's a PEP8 thing but I didn't see it enforced by code_quality so I assumed the codebase uses a different value (we certainly do in ours as 80 is just too small for most modern monitors). If you feel strongly about it lmk and I'll clean it up.

TJStienstra replied:
I do agree with the first point, was just a bit surprised to see this kind of code in tests. Furthermore, thanks for taking the time to update. As for the last point, the 80 character line width is in SymPy as far as I know not a strict rule, but personally I like consistency in formatting (even though I personally prefer 120 chars).

moorepants commented:
> The suggestion sounds reasonable to me (In fact that was the initial https://github.com/sympy/sympy/pull/22818#discussion_r836788688 but we moved away from it)

Sorry to run you in circles. That comment was still tied up with the 2 parts, so maybe not as relevant now. I also made a comment further down about the flag belonging in `.kanes_equations()`. Having a flag and no new attributes does seem cleaner at this point.

zouhairm commented:
Not sure why `test/sphinx` is failing - seems to be unrelated to my changes as it's failing at the installation step:
```
Err:38 http://azure.archive.ubuntu.com/ubuntu focal-updates/main amd64 libgs9-common all 9.50~dfsg-5ubuntu4.5
  404  Not Found [IP: 52.252.75.106 80]
Err:42 http://azure.archive.ubuntu.com/ubuntu focal-updates/main amd64 libgs9 amd64 9.50~dfsg-5ubuntu4.5
  404  Not Found [IP: 52.252.75.106 80]
Err:43 http://azure.archive.ubuntu.com/ubuntu focal-updates/main amd64 ghostscript amd64 9.50~dfsg-5ubuntu4.5
  404  Not Found [IP: 52.252.75.106 80]
Fetched 665 MB in 15s (43.5 MB/s)
E: Failed to fetch http://azure.archive.ubuntu.com/ubuntu/pool/main/g/ghostscript/libgs9-common_9.50~dfsg-5ubuntu4.5_all.deb  404  Not Found [IP: 52.252.75.106 80]
E: Failed to fetch http://azure.archive.ubuntu.com/ubuntu/pool/main/g/ghostscript/libgs9_9.50~dfsg-5ubuntu4.5_amd64.deb  404  Not Found [IP: 52.252.75.106 80]
E: Failed to fetch http://azure.archive.ubuntu.com/ubuntu/pool/main/g/ghostscript/ghostscript_9.50~dfsg-5ubuntu4.5_amd64.deb  404  Not Found [IP: 52.252.75.106 80]
E: Unable to fetch some archives, maybe run apt-get update or try with --fix-missing?
Error: Process completed with exit code 100.
```

TJStienstra commented:
The problem should be fixed in the master now

zouhairm commented on sympy/physics/mechanics/kane.py:
```diff
@@ -117,7 +136,7 @@ class KanesMethod(_Methods):
             configuration_constraints=None, u_dependent=None,
             velocity_constraints=None, acceleration_constraints=None,
-            u_auxiliary=None, bodies=None, forcelist=None):
+            u_auxiliary=None, bodies=None, forcelist=None, explicit_kinematics=True):
```
the way we've implemented it know, explicit_kinematics can be "hot-swapped" as both the implicit and explicit forms are available.
This is because we are only doing the "2nd" part of the original PR of providing the mass/forcing matrices for kinematics in implicit form, i.e. this has no impact on how the dynamics (Kane's) equations themselves are constructed.

I've dropped it from the initializer for now, and we can later add it to the `kanes_equations()` call when the "1" part of the original PR is eventually implemented (i.e. where it can no longer be hotswapped as it would now have an impact on how the dynamics are computed)
I remember the argument for putting it in `kanes_equations()`, which makes sense regarding the "1st" part of the original PR, but since kinematic equations are available before  `kanes_equations()` is invoked, I think it still makes sense to set it in the initial constructor. Perhaps in the future we can decide that `kanes_equations()` can also take it as an argument to overwrite the original value if so desired.


For now I'm tempted to make it a 'private' member so as to signal that it should not be changed mid-run (even though that'd the only way to do it right now if we remove it from the initializer) and to remove it from the initializer and kick the can down the road to figure this out based on how the "1st" part of the original PR is ultimately implemented.

Thoughts?

TJStienstra replied:
Making it a private member is indeed more future safe. As for where we can set the argument, I'm with either of them. `kanes_equations` is probably the most logical location from a users perspective, but `__init__` is the location where we actually need it. So I would prefere making it a private member and putting it in either the `__init__` or `kanes_equations`. Biggest disadvantage I see here is that it is an extra argument, especially in the `__init__` case.

@moorepants what is your opinion?

moorepants replied:
I don't have any issues with adding new arguments to `__init__`. The KanesMethod class does alot already, of course, but it seems in line to let it also return the explicit form of the kinematic equations.

zouhairm replied:
So is the PR fine as is? If so I'll leave to one of you to please merge.
Otherwise please lmk what is left for us to do.

TJStienstra replied:
@moorepants Based on your comments I assumed that you do agree with the current implementation, i.e. `explicit_kinematics` as attribute and as kwarg in the `__init__`, which is also fine with me.
Since I have no more comments on the PR (all suggestions have been taken into account), can this PR be merged?

zouhairm replied:
bump - any chance you can merge please @moorepants ?
Thanks :)

zouhairm committed:
explicit kinematic matrix test

TJStienstra commented on sympy/physics/mechanics/tests/test_kane.py:
```diff
@@ -74,8 +78,20 @@ def test_two_dof():
+
+    # Check that for the implicit form the mass matrix is not identity
+    KM.explicit_kinematics = False
+    assert KM.mass_matrix_kin == Matrix([[1/2, 0], [0, 2]])
```
You probably have to make the `1/2` a sympy object, i.e. `S(1)/2`

zouhairm committed:
fixing SYMENGINE test

moorepants commented:
Merging! Thanks and sorry that we took so long on this.
